main/src/main/java/de/blinkt/openvpn/core/TorProxy.java

```diff
@@ -0,0 +1,316 @@
+            try {
+                inputOpenVPNStream.close();
+            } catch (Exception ignored) {
+
+            }
+            inputOpenVPNStream = null;
+        }
+        if (outputOpenVPNStream != null) {
+            try {
+                outputOpenVPNStream.close();
+            } catch (Exception ignored) {
+
+            }
+            outputOpenVPNStream = null;
+        }
+        isTorConnected = false;
+        pFlow = null;
+        resumeHandler = null;
+    }
+
+    private static void processIncomingPackets(Context c) {
+        ConnectivityManager cm = (ConnectivityManager) c.getSystemService(CONNECTIVITY_SERVICE);
+        PackageManager pm = c.getPackageManager();
+        byte[] buf = new byte[16384];
+        while (!Thread.interrupted()) {
+            try {
+                int read = inputOpenVPNStream.read(buf);
+                if (read > 0) {
+                    int protocol = getProtocol(buf);
+                    if (protocol == 6 || protocol == 17) {
+                        InetSocketAddress sourceAddress = getSourceAddress(buf, protocol);
+                        InetSocketAddress destinationAddress = getDestinationAddress(buf, protocol);
+                        int uid = cm.getConnectionOwnerUid(protocol, destinationAddress, sourceAddress);
+                        String appName = KNOWN_UIDS.computeIfAbsent(uid, pm::getNameForUid);
+                        if (OpenVPNService.ORBOT_PACKAGE_NAME.equals(appName)) {
+                            outputDeviceStream.write(buf, 0, read);
+                        }
+                    }
+                }
+            } catch (Exception e) {
+                Log.e(TAG, "Error while transferring packets from device to vpn", e);
+            }
+        }
+    }
+
+    private static void sendPacketsToTor(byte[] buf, int read) {
+        try {
+            if (read > 0) {
+                byte[] data = Arrays.copyOf(buf, read);
+                try {
+                    Packet packet = IpSelector.newPacket(data, 0, data.length);
+                    if (packet instanceof IpPacket) {
+                        IpPacket ipPacket = (IpPacket) packet;
+                        if (isPacketDNS(ipPacket)) {
+                            executorService.execute(new RequestPacketHandler(ipPacket, pFlow, dnsResolver));
+                        } else {
+                            IPtProxy.inputPacket(data);
+                        }
+                    }
+                } catch (IllegalRawDataException e) {
+                    Log.e(TAG, e.getLocalizedMessage());
+                }
+            }
+        } catch (Exception e) {
+            Log.d(TAG, "error reading from VPN fd: " + e.getLocalizedMessage());
+        }
+    }
+
+    private static boolean isPacketDNS(IpPacket p) {
+        if (p.getHeader().getProtocol() == IpNumber.UDP) {
+            UdpPacket up = (UdpPacket) p.getPayload();
+            return up.getHeader().getDstPort() == UdpPort.DOMAIN;
```

Comment:
This sounds like it might break/throw an expection if you have an invalid packet.